### PR TITLE
Change shape of quadrature tensors for better broadcasting

### PR DESCRIPTION
**Pull request by gustavocmv**
**PR type:** Enhancement

**Related issue(s)/PRs:**  #1505

## Summary

**Proposed changes**

PR #1505 introduced a new approach for quadrature in the form of classes. In the original code, the penultimate dimension was used to broadcast with the quadrature points, resulting in applying the "quadratured" function f over tensors with shape `[N1, ..., Nb, N_gh_points, Dim]`. This resulted in awkward broadcasting  of the quadrature-tensors with other tensors. To change it, the quadrature code was modified to use  shape `[ N_gh_points, N1, ..., Nb, Dim]`

**What alternatives have you considered?**

I wanted to have it working that way when I worked with #1505, but couldn't make it work.

## PR checklist

- [ ] New features: code is well-documented
  - [ ] detailed docstrings (API documentation)
  - [ ] notebook examples (usage demonstration)
- [ ] The bug case / new feature is covered by unit tests
- [X] Code has type annotations
- [X] I ran the black formatter (`make format`)
- [X] I locally tested that the tests pass (`make check-all`)

### CODE REVIEW

File: gpflow/quadrature/base.py [hunk 1]
```diff
@@ -31,26 +31,26 @@ def __call__(self, fun, mean, var, *args, **kwargs):
...
         :param fun: Callable or Iterable of Callables that operates elementwise, with
-            signature f(X, *args, **kwargs). Moreover, if must satisfy the shape-mapping:
-                X shape: [b1, b2, ..., bX, N_quad_points, d],
-                    usually [N, N_quad_points, d]
-                f(X) shape: [b1, b2, ...., bf, N_quad_points, d'],
-                    usually [N, N_quad_points, 1] or [N, N_quad_points, d]
+            signature f(X, *args, **kwargs). Moreover, it must satisfy the shape-mapping:
+                X shape: [N_quad_points, b1, b2, ..., bX, d],
+                    usually [N_quad_points, N, d]
+                f(X) shape: [N_quad_points, b1, b2, ...., bf, d'],
```
Comment (st--) [inline gpflow/quadrature/base.py:37]: do `bf` and `bX` have to be the same? I'm assuming `d` and `d'` can be different?

File: gpflow/quadrature/gauss_hermite.py [hunk 2]
```diff
@@ -91,20 +91,26 @@ def _build_X_W(self, mean: TensorType, var: TensorType):
...
-        W = self.dZ
-        # X: [b1, b2, ..., bX, n_gh_total, dim], usually [N, n_gh_total, dim]
-        # W: [b1, b2, ..., bX, n_gh_total,   1], usually [N, n_gh_total,   1]
+        # mean, var: [b1, b2, ..., bX, dim], usually [N, dim]
+        mean = tf.expand_dims(mean, 0)
+        stddev = tf.expand_dims(tf.sqrt(var), 0)
+        # mean, stddev: [1, b1, b2, ..., bX, dim], usually [1, N, dim]
+
+        Z = tf.reshape(self.Z, tf.concat([shape_aux, [self.dim]], axis=0))
+        dZ = tf.reshape(self.dZ, tf.concat([shape_aux, [1,]], axis=0))
```
Comment (st--) [inline gpflow/quadrature/gauss_hermite.py:109]: what was your reason for not just changing the shapes of `Z` and `dZ` in the `__init__`?

File: gpflow/quadrature/base.py (hunk 1 above)
Comment (gustavocmv) [reply gpflow/quadrature/base.py:37]: `bf` and `bX` should be the same -- I will fix it.
`d` and `d'` can be different.

File: gpflow/quadrature/gauss_hermite.py (hunk 2 above)
Comment (gustavocmv) [reply gpflow/quadrature/gauss_hermite.py:109]: `shape_aux` depends on the shape of `mean`, which is available only when calling `_build_X_W(mean, var)`

File: gpflow/quadrature/deprecated.py [hunk 3]
```diff
@@ -136,7 +136,7 @@ def ndiagquad(funcs, H: int, Fmu, Fvar, logspace: bool = False, **Ys):
 
     def wrapper(old_fun):
         def new_fun(X, **Ys):
-            Xs = tf.unstack(X, axis=-1)
+            Xs = tf.unstack(tf.expand_dims(X, axis=-2), axis=-1)
```
Comment (st--) [inline gpflow/quadrature/deprecated.py:139]: why this change ?

Comment (gustavocmv) [reply gpflow/quadrature/deprecated.py:139]: This was necessary to make the wrapper work with the new shape.

**Review by vdutor — APPROVED** [review]
